### PR TITLE
Attempt to time-out potentially stalled watches

### PR DESCRIPTION
**Pull request by willscott**
potentially related to #803

### CODE REVIEW

File: http/server.go [hunk 1]
```diff
@@ -101,16 +101,30 @@ func (h *handler) start() {
...
+	h.chainInfoLk.RUnlock()
 	for {
-		next, ok := <-stream
+		var next client.Result
+		var ok bool
+		select {
+		case next, ok = <-stream:
+		case <-time.After(expectedRoundDelayBackoff):
+			cncl()
+			goto RESET
```
Comment (lukevalenta) [inline http/server.go:126]: I'd love to get rid of this `goto`. Maybe we could use a helper function? Something like
```
func (h *handler) Watch(ctx context.Context, ready chan bool) {
    for {
        select {
        case <-ctx.Done():
            return
        default:
        }
        h.watchWithTimeout(ctx, ready)
    }
}

func (h *handler) watchWithTimeout(ctx context.Context, ready chan bool) {
    watchCtx, cncl := context.WithCancel(ctx)
    stream := h.client.Watch(watchCtx)
    for {
        var next client.Result
	var ok bool
	select {
        case <-ctx.Done():
            return
	case next, ok = <-stream:
	case <-time.After(expectedRoundDelayBackoff):
             return
	}
	if !ok {
	    h.log.Warn("http_server", "random stream round failed")
	    h.pendingLk.Lock()
            h.latestRound = 0
	    h.pendingLk.Unlock()
	    // backoff on failures a bit to not fall into a tight loop.
	    // TODO: tuning.
	    time.Sleep(watchConnectBackoff)
	    return
         }
         ...
    }
}
```

File: http/server.go [hunk 2]
```diff
@@ -141,6 +155,7 @@ RESET:
 
 		select {
 		case <-ctx.Done():
```
Comment (lukevalenta) [inline http/server.go:157]: Could we move this `ctx.Done()` to the new `select` at https://github.com/drand/drand/pull/843/files#diff-9cdb0ee2b48996738ec1fc3037630e10eb388ebc30fb400df2e627590f0148c6R122?

**Comment by willscott** [comment]
agree that's cleaner

**Review by lukevalenta — APPROVED** [review]
Looks good to me! Thanks for making the change.

**Pull request merged**